webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
+                for index in range {
+                    let image = &images[&bound_textures[index]].core;
+                    // We need to transit the image, even though it's bound in the descriptor set
+                    let mut src_stage = Some(hal::pso::PipelineStage::empty());
+                    if let Some(barrier) = image.transit(
+                        hal::image::Access::SHADER_READ,
+                        hal::image::Layout::ShaderReadOnlyOptimal,
+                        image.subresource_range.clone(),
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
+                                src_stage.unwrap()
+                                    .. hal::pso::PipelineStage::FRAGMENT_SHADER,
+                                hal::memory::Dependencies::empty(),
+                                &[barrier],
+                            );
+                        }
+                    }
                 }
-                &mut self.cache_clip_pool[location.pool_idx]
             },
-            ShaderGroup::Brush => {
-                if location.pool_idx != self.brush_pool_idx
-                    && !self.marked_brush_pools.contains(&location.pool_idx) {
-                    self.marked_brush_pools.push(location.pool_idx)
-                }
-                &mut self.brush_pool[location.pool_idx]
+            Entry::Vacant(v) => {
+                let free_sets = self.free_sets.get_mut(group);
+                let desc_set = match free_sets.pop() {
+                    Some(ds) => ds,
+                    None => {
+                        unsafe {
+                            desc_allocator.allocate(
+                                device,
+                                group_data.descriptor_layout(group, set_index),
+                                group_data.ranges(group, set_index),
+                                DESCRIPTOR_COUNT,
+                                free_sets,
+                            )
+                        }.expect("Allocate descriptor sets failed");
+                        free_sets.pop().unwrap()
+                    }
+                };
+                let desc_set = v.insert(desc_set);
+                let descriptor_writes = RENDERER_TEXTURE_BINDINGS[range.clone()].iter().zip(range.into_iter()).map(|(binding, index)| {
+                    let image = &images[&bound_textures[index]].core;
+                    let mut src_stage = Some(hal::pso::PipelineStage::empty());
+                    if let Some(barrier) = image.transit(
+                        hal::image::Access::SHADER_READ,
+                        hal::image::Layout::ShaderReadOnlyOptimal,
+                        image.subresource_range.clone(),
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
```

Comment:
similarly, would be good to combine these barriers into a single one.
All in all, maybe we can move the barrier generation out of the `match` arms, since it should be similar for both?